plugins/parsers/json/parser_test.go

```diff
@@ -448,22 +449,148 @@ func TestJSONParseNestedArray(t *testing.T) {
 	"total_devices": 5,
 	"total_threads": 10,
 	"shares": {
-	"total": 5,
-	"accepted": 5,
-	"rejected": 0,
-	"avg_find_time": 4,
-	"tester": "work",
-	"tester2": "don't want this",
-	"tester3": 7.93
+		"total": 5,
+		"accepted": 5,
+		"rejected": 0,
+		"avg_find_time": 4,
+		"tester": "work",
+		"tester2": "don't want this",
+		"tester3": {
+			"hello":"sup",
+			"fun":"money",
+			"break":9
+		}
 	}
 	}`
 
 	parser := JSONParser{
 		MetricName: "json_test",
-		TagKeys:    []string{"total_devices", "total_threads", "shares_tester", "shares_tester3"},
+		TagKeys:    []string{"total_devices", "total_threads", "shares_tester3_fun"},
 	}
 
 	metrics, err := parser.Parse([]byte(testString))
+	log.Printf("m[0] name: %v, tags: %v, fields: %v", metrics[0].Name(), metrics[0].Tags(), metrics[0].Fields())
 	require.NoError(t, err)
 	require.Equal(t, len(parser.TagKeys), len(metrics[0].Tags()))
 }
+
+func TestJSONQueryErrorOnArray(t *testing.T) {
+	testString := `{
+		"total_devices": 5,
+		"total_threads": 10,
+		"shares": {
+			"total": 5,
+			"accepted": 6,
+			"test_string": "don't want this",
+			"test_obj": {
+				"hello":"sup",
+				"fun":"money",
+				"break":9
+			},
+			"myArr":[4,5,6]
+		}
+	}`
+
+	parser := JSONParser{
+		MetricName: "json_test",
+		TagKeys:    []string{},
+		JSONQuery:  "shares.myArr",
+	}
+
+	_, err := parser.Parse([]byte(testString))
+	assert.Error(t, err)
+}
+
+func TestArrayOfObjects(t *testing.T) {
+	testString := `{
+		"meta": {
+			"info":9,
+			"shares": [{
+				"channel": 6,
+				"time": 1130,
+				"ice":"man"
+			},
+			{
+				"channel": 5,
+				"time": 1030,
+				"ice":"bucket"
+			},
+			{
+				"channel": 10,
+				"time": 330,
+				"ice":"cream"
+			}]
+		},
+		"more_stuff":"junk"
+	}`
+
+	parser := JSONParser{
+		MetricName: "json_test",
+		TagKeys:    []string{"ice"},
+		JSONQuery:  "meta.shares",
+	}
+
+	metrics, err := parser.Parse([]byte(testString))
+	assert.NoError(t, err)
+	assert.Equal(t, 3, len(metrics))
+}
+
+func TestUseCaseJSONQuery(t *testing.T) {
+	testString := `{
+		"obj": {
+			"name": {"first": "Tom", "last": "Anderson"},
+			"age":37,
+			"children": ["Sara","Alex","Jack"],
+			"fav.movie": "Deer Hunter",
+			"friends": [
+				{"first": "Dale", "last": "Murphy", "age": 44},
+				{"first": "Roger", "last": "Craig", "age": 68},
+				{"first": "Jane", "last": "Murphy", "age": 47}
+			]
+		}
+	}`
+
+	parser := JSONParser{
+		MetricName:   "json_test",
+		StringFields: []string{"last"},
+		TagKeys:      []string{"first"},
+		JSONQuery:    "obj.friends",
+	}
+
+	metrics, err := parser.Parse([]byte(testString))
+	assert.NoError(t, err)
+	assert.Equal(t, metrics[0].Fields()["last"], "Murphy")
+	assert.Equal(t, 3, len(metrics))
```

Comment:
Check the length before indexing, this will make the test easier to understand if it fails.